### ISSUE TITLE
SSL_ERROR_SSL: error:1408F10B:..routines:SSL3_GET_RECORD:wrong version ..

### ISSUE THREAD

**Issue by git123456** [IBM-Blockchain/ibm-blockchain-issues#32]
Hi,

Today I just updated my membership and peer docker images, installed new hfc (v0.6.0), copied the membership image certificate, tlsca.cert to my local test environment.  Now trying to run the hfc simple test program (chain_test), attached, to register new user, but got the following errors :+1: 
E0912 17:05:29.361932000 123145321250816 ssl_transport_security.c:945] Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number.
E0912 17:05:29.361963000 123145321250816 handshake.c:128] Security handshake failed: {"created":"@1473714329.361950000","description":"Handshake failed","file":"../src/core/lib/security/transport/handshake.c","file_line":264,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
E0912 17:05:29.361986000 123145321250816 secure_channel_create.c:98] Secure handshake failed with error 1.
I0912 17:05:29.362022000 123145321250816 subchannel.c:642] Connect failed: null
I0912 17:05:29.362029000 123145321250816 subchannel.c:647] Retry in 0.998263936 seconds
E0912 17:05:29.363362000 123145321250816 ssl_transport_security.c:945] Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number.
E0912 17:05:29.363400000 123145321250816 handshake.c:128] Security handshake failed: {"created":"@1473714329.363375000","description":"Handshake failed","file":"../src/core/lib/security/transport/handshake.c","file_line":264,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
E0912 17:05:29.363421000 123145321250816 secure_channel_create.c:98] Secure handshake failed with error 1.
I0912 17:05:29.363446000 123145321250816 subchannel.c:642] Connect failed: null
I0912 17:05:29.363451000 123145321250816 subchannel.c:647] Retry in 0.998592899 seconds

any idea?

**Comment by angrbrd** [comment]
Hi @git123456 

I am working to update hfc v0.6.x in the next few days to address tls related issues. I believe it should resolve what you are seeing. Please stay tuned for an upcoming update.

Thanks!

**Comment by git123456** [comment]
Hi,

Thanks for the update, wasted one day trying to debug this problem.  Two more questions …
1) In general, where or how do I know whenever there's an hfs update?

2) I may have similar problem with the bluemix one, question is how do I check what version of hfc bluemix is using?

**Comment by angrbrd** [comment]
Hi @git123456,

1) I am not aware of a way to "subscribe" to receive an update on a new package. You just specify in your `package.json` file that you would like to set a specific hfc version. For example, if you specifically need to use version 0.5.3 (against Bluemix) you would include:

"hfc": "0.5.3"

If you need anything above 0.6.0, you would include:

"hfc": "^0.6.0" 

And it will automatically update your package when you do an `npm install`.

2) You should use hfc 0.5.3 against Bluemix. But Your problem in issue#31 is not related to the TLS issue you are observing here.

Thanks.

**Comment by git123456** [comment]
Hi angrgbird,

Yes I do know how to use package.json, but that was not my question. If I don't know what version existed and what to use with what then I wouldn't know what to put in package.json, right?  This is simply FYI.

**Comment by dshuffma-ibm** [comment]
_2) I may have similar problem with the bluemix one, question is how do I check what version of hfc bluemix is using?_

well bluemix has little to do with HFC (the sdk), but if you mean which version of Hyperledger (the fabric) then you can look it up via the "Status" tab.  Look by the "Release Notes" and find the disclaimer that says "Your network is using this version". Actually I apologize here, there were dead links because some smart people decided to move stuff around, But I've fixed them. Click the hyperledger commit and you can see its from branch **v0.5-developer-preview**

![image](https://cloud.githubusercontent.com/assets/16085295/18489408/651c141a-79cb-11e6-8e62-226fcf9803f4.png)

tldr: Bluemix uses Hyperledger **v0.5-developer-preview** and you should use **hfc v0.5.**\*

**Comment by angrbrd** [comment]
@git123456, Not sure which exact information will help you here, so I'll just mention several things. If you are looking to find out which npm versions of hfc are already published you could issue the following command.

`npm show hfc@* version`
hfc@0.0.1 '0.0.1'
hfc@0.0.2 '0.0.2'
hfc@0.0.3 '0.0.3'
hfc@0.0.4 '0.0.4'
hfc@0.0.5 '0.0.5'
hfc@0.5.0 '0.5.0'
hfc@0.6.0 '0.6.0'
hfc@0.5.1 '0.5.1'
hfc@0.5.2 '0.5.2'
hfc@0.5.3 '0.5.3'

If you are wondering specifically on compatibility, I have updated the README to clarify that:
https://www.npmjs.com/package/hfc#installation

@dshuffma-ibm, Bluemix does have to do with the SDK as it is running a very specific version of Hyperledger Fabric. Therefore, if you want to use the SDK with a Bluemix network, you have to use a specific version of hfc as well, which is 0.5.3. For example version 0.6.0 will definitely not work against Bluemix.

Thanks!

**Comment by mrshah-at-work** [comment]
@angrbrd do you suggest that there should be a _Supported HFC Version_ in the release notes on bluemix? If yes, what should it be for fabric `v0.5`?

**Comment by git123456** [comment]
thanks angrdbird and dshuffma, you both do provide some of the versioning information I was looking but still missing one piece of information which is how do I know what hfc version works with or has been tested with what version of fabric. My bluemix blkchain, in the release notes it stated that the commit level is V0.4.1. then my question is how do I know what version of hfc works with fabric V0.4.1?

**Comment by dshuffma-ibm** [comment]
@git123456 looks like angrbrd updated the npm readme with the information you are after: 

![image](https://cloud.githubusercontent.com/assets/16085295/18514964/928c3592-7a61-11e6-8494-c53840bab61a.png)

https://www.npmjs.com/package/hfc#installation

I think it makes the most sense for this information to live here too instead of addding it to the bluemix service release notes @mrshah-at-ibm

**Comment by git123456** [comment]
Hi Dshuffma, yes I did read what angrybird has updated but my bluemix fabric is v0.4.1. then what version of hfc should I use?  What he updated only targeted to specific fabric version which is 0.5 (right?)

**Comment by git123456** [comment]
@angrbrd interesting... when I did `npm show hfc@* version` , it shows as follows though
[
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {},
  {}
]

any idea?
(this is on Mac)

**Comment by angrbrd** [comment]
@git123456, I am also on a Mac so not sure what to tell you immediately. Check your npm version or see if there is another npm command that would return the same info. This is what I have in my environment.

```
$ npm version
{ npm: '2.14.20',
  ares: '1.10.1-DEV',
  http_parser: '2.5.2',
  icu: '56.1',
  modules: '46',
  node: '4.4.0',
  openssl: '1.0.2g',
  uv: '1.8.0',
  v8: '4.5.103.35',
  zlib: '1.2.8' }
$ npm show hfc@* version
hfc@0.0.1 '0.0.1'
hfc@0.0.2 '0.0.2'
hfc@0.0.3 '0.0.3'
hfc@0.0.4 '0.0.4'
hfc@0.0.5 '0.0.5'
hfc@0.5.0 '0.5.0'
hfc@0.6.0 '0.6.0'
hfc@0.5.1 '0.5.1'
hfc@0.5.2 '0.5.2'
hfc@0.5.3 '0.5.3'
```

**Comment by dshuffma-ibm** [comment]
So lets get this versioning thing straight b/c there are lots of things each with their own version number.
1. Your bluemix service is **IBM Blockchain Service v0.4.1**
2. From the bluemix service version we can tell the network is using **Hyperledger Fabric branch v0.5-developer-preview** by looking at the the release notes (click the commit url, look at the branch name in github)
3. Using that information we can look at [HFC's readme](https://www.npmjs.com/package/hfc#installation) and see that **HFC v0.5.3** is the latest for this fabric version

**Comment by git123456** [comment]
@angrbrd any idea when the 'hfc v0.6.x' which should fix this problem be released?  Or an estimation of date would help.  I may be out of network connection for the next few days, can really use this fix to continue working.

**Comment by angrbrd** [comment]
@git123456 we are testing the changes today. I am hoping to publish by end of today or if not today then tomorrow for sure.

**Comment by git123456** [comment]
@angrbrd I see that there is hfc 0.6.1, I tried it but got different errors as shown below.  BTW, do I have to update my docker images in order to use this new hfc ?

E0919 11:57:39.488635000 123145338589184 ssl_transport_security.c:877] Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number.
E0919 11:57:39.488819000 123145338589184 handshake.c:241] Handshake failed with error TSI_PROTOCOL_FAILURE
E0919 11:57:39.488829000 123145338589184 secure_channel_create.c:102] Secure handshake failed with error 1.
E0919 11:57:39.489883000 123145338589184 ssl_transport_security.c:877] Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number.
E0919 11:57:39.489896000 123145338589184 handshake.c:241] Handshake failed with error TSI_PROTOCOL_FAILURE
E0919 11:57:39.489902000 123145338589184 secure_channel_create.c:102] Secure handshake failed with error 1.
E0919 11:57:40.492482000 123145338589184 ssl_transport_security.c:877] Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number.

**Comment by git123456** [comment]
@angrbrd I updated the docker images anyway, use the original tlsca.cert and also tried the fixtures folder's tlsca.cert, now the error is back to the original SSL error :+1: 

E0919 13:01:54.514576000 123145313386496 ssl_transport_security.c:877] Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number.
E0919 13:01:54.514798000 123145313386496 handshake.c:241] Handshake failed with error TSI_PROTOCOL_FAILURE
E0919 13:01:54.514808000 123145313386496 secure_channel_create.c:102] Secure handshake failed with error 1.
E0919 13:01:54.516024000 123145313386496 ssl_transport_security.c:877] Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number.

any idea?  Why is it so hard to get this simple thing to run?

**Comment by git123456** [comment]
@angrbrd @dshuffma-ibm @mrshah-at-ibm could some one response, a simple note 'not able to solve it, good luck!!!' is good enough...  I am becoming real angry bird dealing with this persistent problem (thought this simple scenario logically should be one of your simple test case.. guess not)

**Comment by masterDev1985** [comment]
This may or may not be useful to you, but I found a thread that mentions a method for debugging this error message:
https://groups.google.com/forum/#!topic/git-users/5cQ4I7qRx0I
See the first response from Konstantin.  @git123456 , it might we worth trying the `openssl s_client` command against the peer endpoint that hfc is failing to connect to.  That will tell you what versions of TLS that endpoint supports.  After that, we might need to look at the hfc code to tell what version of TLS it tries to connect with.

**Comment by angrbrd** [comment]
@git123456 

We understand that you are quite frustrated with this issue and we are sorry about that. However, please understand, that we are working to improve this nascent project and things are changing, getting better, and more stable as we move forward. Please have patience, as this is certainly not production ready code and expecting this to be production ready with absolutely no issues is simply the wrong expectation to have. This is why we invite the community to contribute and work with us, as well as to open up issues (like you have done), or suggest new features to continuously improve this project. This being said, it appears to me, that you are not setting up TLS correctly. This likely implies that our documentation describing TLS configuration (_Enabling TLS_ section at the link below) is not good enough to explain what must be done to get it working right.

http://hyperledger-fabric.readthedocs.io/en/latest/nodeSDK/node-sdk-indepth/

We have someone on the team working to reproduce the problem right now so we can update the documentation accordingly to have a clear understanding of how to get TLS setup working. Thank you for your patience.

**Comment by rennman** [comment]
@git123456 @angrbrd 

At least one scenario exists in which this exact same message is issued in the GRPC_TRACE:
"Handshake failed with fatal error SSL_ERROR_SSL: error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number."
This is generated when an sdk client attempts to communicate over a secure socket (i.e., negotiate TLS), but the remote target of the connection is not enabled for TLS. 

This raises the possibility that one or more entities in the network are not configured for TLS. This could happen for instance if you had a error in the *yaml file or syntax error in an environment variable, e.g.
      MEMBERSRVC_CA_SERVER_TLS_CERTFILE
vs.
      MEMBERSRVC_CA_SERVER_TLS_CERT_FILE
(as an aside, note that the cited example is precisely one of the differences between the v0.5 branch and the v0.6)

Are you exclusively using the *yaml files to feed the configuration to the servers? Or some combination of that and ENVVARs?

It would be very helpful if you could attach the relevant  membersrvc.yaml and core.yaml files, in addition to any environmental customizations you may have.

**Comment by git123456** [comment]
@angrbrd @rennman I thought I was pretty patient...the fact that you mentioned the new hfs version would fix this SSL error but it was not the case when I tried it (v0.6.1) hence I was confuse and lost.

Yes the doc regarding needed to set TLS was not clear, also the optional steps needed to reconfigured docker images (to enable TLS???) were not listed as well.

Questions :
1) how do I know if I need to enable TLS?
2) Also in the Enabling TLS doc (the link that you mentioned), pls clarify the following steps :
    - the changes needed in $GOPATH/src/github.com/hyperledger/fabric/membersrvc/membersrvc.yaml,  I am assuming this is in membership node, right?
    - the changes needed in $GOPATH/src/github.com/hyperledger/fabric/peer/core.yaml, is this in peer or membership node?
    - the statement "Copy /var/hyperledger/production/.membersrvc/tlsca.cert to $GOPATH/src/github.com/hyperledger/fabric/sdk/node", where is this? In membership. peer or client node?
3) Does tlse.cert get generated a new one each time membership get restarted again (note I am using docker images)?  Wondering if I have to keep coping the tlse.cert to the client site each time I restarted the membership docker image.

Thanks for you understanding.

note : I didn't customize anything at all, this is the latest published docker images running locally.

**Comment by rennman** [comment]
@angrbrd @git123456 

1) TLS is enabled in IBM-supplied networks, including the Bluemix starter and HSBN networks. Hence if you are using these services, TLS is required. If running locally, or in a network that you setup on your own, TLS is optional.

3 ) 
2 and 3) are related; I'll answer them in reverse. All of the files in the /var/hyperledger/production/.membersrvc/ directory (including the server's public/private key pair and associated X509v3 certificate) are dynamically generated when membersrvc starts. Therefore, the sequence should be
1. start membersrvc
2. copy the sever certificate, /var/hyperledger/production/.membersrvc/tlsca.cert

Note that, once created, the data in this directory is persistent across multiple re-starts of membersrvc. If the directory is deleted (e.g. between two disjoint set of tests), it will be re-created from scratch and populated with an initial set of data. Any time the certificate is re-generated, it will need to be copied again.

The directory is on the filesystem where membersrvc is running:
1. if running on a native platform, it is on the local FS
2. if running in vagrant, it is on the guest VM's FS
3. if running in docker, it is on the docker container's FS

e.g. in case 3, 
docker cp ${membersrvc_container_ID}:/var/hyperledger/production/.membersrvc/tlsca.cert $GOPATH/src/github.com/hyperledger/fabric/sdk/node

The current SDK scheme assumes you will run the tests using $GOPATH/src/github.com/hyperledger/fabric/sdk/node as your CWD, since it checks for the existence of the tlsca.cert file at that location in order to decide to use TLS. (This mechanism can and will be improved.)

Yes, 
$GOPATH/src/github.com/hyperledger/fabric/membersrvc/membersrvc.yaml controls the settings for the membersrvc daemon (in some contexts also referred to as the 'CA')
while
$GOPATH/src/github.com/hyperledger/fabric/peer/core.yaml controls the settings for the peer node.

Nearly all of the settings in either of these 2 files can be explicitly overridden using environment variables. membersrvc variables use the prefix "MEMBERSRVC_CA_", while the peer variables use the prefix "CORE_"

The relevant ones for TLS would be:
MEMBERSRVC_
v0.6
- export MEMBERSRVC_CA_SERVER_TLS_CERT_FILE=$SDK_CA_CERT_FILE
- export MEMBERSRVC_CA_SERVER_TLS_KEY_FILE=$SDK_CA_KEY_FILE

v0.5
- export MEMBERSRVC_CA_SERVER_TLS_CERTFILE=$SDK_CA_CERT_FILE
- export MEMBERSRVC_CA_SERVER_TLS_KEYFILE=$SDK_CA_KEY_FILE

This not only identifies the certificate/key pair, it also implicitly enables TLS. There is no explicit 'TLS' flag for membersrvc

PEER_
- export CORE_PEER_TLS_ENABLED=true
- export CORE_PEER_TLS_CERT_FILE=$SDK_CA_CERT_FILE
- export CORE_PEER_TLS_KEY_FILE=$SDK_CA_KEY_FILE
- export CORE_PEER_TLS_SERVERHOSTOVERRIDE=tlsca
- export CORE_PEER_PKI_TLS_ENABLED=true
- export CORE_PEER_PKI_TLS_ROOTCERT_FILE=$SDK_CA_CERT_FILE
- export CORE_PEER_PKI_TLS_SERVERHOSTOVERRIDE=tlsca`

serverhostoverride indicates that the identity included in the remote partner's certificate does not necessarily have to match the address or URL of the request, thereby bypassing strict identify enforcement  - a common TLS check and enabled in the underlying GRPC module. If running locally, you could also choose to omit this parameter by adding an entry to /etc/hosts

```
   127.0.0.1    localhost tlsca
```

By using the latter alternative, you would set the corresponding URL for the SDK to
grpcs://tlsca:${listening_port}
e.g.
chain.setMemberServicesUrl("grpc://tlsca:7054")

**Comment by git123456** [comment]
@angrbrd I gave up using docker image... I will just stick with bluemix.  
BTW, I accidentally pull the latest fabric latest version,now trying to get  fabric v0.5-developer-preview-1 as you mentioned in https://www.npmjs.com/package/hfc#installation ( I am assuming you meant the fabric codes)
but was not able to do so.. I may not be doing it correctly, this is what I did 'git clone https://github.com/hyperledger/fabric.git -b v0.5-developer-preview-1 --single-branch fabric_05', but it failed, can you pls advice. 

p/s no problem to install hfc@0.5.3

**Comment by angrbrd** [comment]
@git123456 

If you are working only with Bluemix, you do not need to clone the Fabric v0.5-developer-preview branch. You just need to have a blockchain network deployed in Bluemix and the hfc v0.5.3 installed in order to interact with the network deployed in Bluemix. 

However, if you are looking to spin up your own network (outside of Bluemix) using a _newer_ version of Fabric, then yes you would need to clone Fabric at the link below:

https://github.com/hyperledger/fabric

Perhaps, you can better explain to us what you are trying to accomplish and myself, @dshuffma-ibm @mrshah-at-ibm @masterDev1985 and @rennman can better help you get there. If you are looking to just get started to understand the functionality, the Bluemix service is definitely the way to go. Guys, please chime in if you have a better place to start.

Thanks.

**Comment by mrshah-at-work** [comment]
If you want to get v0.5-developer-preview, it was moved to https://github.com/hyperledger-archives/fabric/tree/v0.5-developer-preview when the project was moved to used Gerrit.
Thus, it failed to clone from the hyperledger/fabric repository.

**Comment by git123456** [comment]
@angrbrd @dshuffma-ibm @mrshah-at-ibm Ideally I want to have (1) a blockchain network running locally (hence the docker images) because I may not have access to network for some time but still need to work on blockchain (2) a bluemix blockchain, coming from big data background I know program that works in one node network may not work in >1 node networks hence bluemix blockchain is more for sanity testing.

For now I am still not able to get the docker images to work locally hence I am stuck with bluemix blockchain.

The reason I asked for Fabric v0.5-developer-preview was because I thought, correctly me if I am wrong, the shim interface for 0.5 version seem to be different from the 0.6.1 version which I accidentally updated and wipe up my older fabric version.  I was told bluemix works with hfc 0.5.3 version (which implied fabric 0.5 version??) but as stated my older fabric version was accidentally replaced by 0.6 version, now I need to deploy a chain code in bluemix but was not able to do so (with different version of shim?).  Now let me know if I may have misunderstood anything here.

**Comment by angrbrd** [comment]
@git123456 

(1) If you need to test locally you can do so via two different means: (a) start up a single peer and a membership service process. Yes, that only gives you a single instance of each, but that is the easiest way to test things out quickly. You can use this simple approach without TLS to quickly get started and develop/test some chaincode. (b) start up peer(s) and a membership service in Docker containers. You can also use this approach without TLS. And as TLS has frustrated you by now, I suggest you give it a shot without TLS while we modify our documentation to remove any TLS related ambiguities.

As for the latter part of your question, yes I believe you have the correct understanding regarding the v0.5 being incompatible with the v0.6. @mrshah-at-ibm and @dshuffma-ibm please clarify.

**Comment by git123456** [comment]
@angrbrd @dshuffma-ibm  ... now I am back to trying to get this unstable docker images to work because now my bluemix failed to deploy new chain code (issue 39) ... very frustrated with this... this is suppose to be a very simple process...unbelievable!

BTW, all this time I was testing with 1 peer only and starting the membership and 1 peer with the attached docker-compos file, do verify if anything wrong it.

Pls answer the following questions :
1. You did say some one would try to test it using docker images.  Can you tell me what was the result and what version of docker images he has tried?

2.Is it true that if I don't use SSL (hence using grpc instead of grpcs) then I don't have to worry about any kid of certificate (tlsca.cert?), right?

3) Starting the peer images, what does "CORE_SECURITY_ENABLED=false" implied?  Means I don't have to use grpcS?  Did try both true and false.
[docker-compose.txt](https://github.com/IBM-Blockchain/ibm-blockchain-issues/files/489654/docker-compose.txt)

**Comment by ratnakar-asara** [comment]
@git123456 
See if this little program is of any help
https://github.com/ratnakar-asara/NodeSdkSample/tree/master/tlsca

**Comment by mrshah-at-work** [comment]
We have updated the service to v0.6 of fabric on Nov 14 and hfc v0.6.5 should work with it. Also, the images and docker-compose have been published for local deployments at https://hub.docker.com/r/ibmblockchain/fabric-peer/

Please let us know if you still see the issue. Closing this for now.
